File: LC/lc144/lc144.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
  let res = [];
  let helper = function(t){
      if (!t){return;} else{
          res.push(t.val);
          helper(t.left);
          helper(t.right);
          return;
      }
  }
  helper(root);
  return res;
};

File: LC/lc144/lc144.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> pre;
    stack<TreeNode*> s;
    void traverse(TreeNode* t) {
        s.push(t);
        while(!s.empty()){
            TreeNode* n = s.top();
            s.pop();
            if (n != nullptr){
                pre.push_back(n->val);
                s.push(n->right);
                s.push(n->left);
            }
        }
    }
    vector<int> preorderTraversal(TreeNode* root) {
        traverse(root);
        return pre;
    }
};